File: LeetCode/111-Minimum-Depth-of-Binary-Tree(Easy)/Solution2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 * 
 * Iterative DFS:
 * Time O(n)
 * Space O(n) (worst, completely unbalanced)
 *       O(logn) (best, completely balanced)
 */
class Solution {
    public int minDepth(TreeNode root) {
        if (root == null)
            return 0;
        
        int depth = 1;
        int minDepth = Integer.MAX_VALUE;
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);

        while(!stack.isEmpty()) {
            TreeNode peekNode = stack.peek();

            if (peekNode.left != null) {
                stack.push(peekNode.left);
            }

            if (peekNode.right != null) {
                stack.push(peekNode.right);
            }

            if (peekNode.left == null && peekNode.right == null) {
                minDepth = Math.min(depth, minDepth);
                // go back to the branch      
                TreeNode popNode = stack.pop();
                while (popNode != null && !stack.isEmpty()) {
                    peekNode = stack.peek();
                    if (isParentChild(peekNode, popNode)) {
                        --depth;
                        popNode = stack.pop();
                    } else {
                        break;
                    }                    
                }
            } else {
                ++depth;
            }
        }
        return minDepth;
    }

    private boolean isParentChild(TreeNode parent, TreeNode child) {
        return (child == parent.left || child == parent.right);
    }
}

File: LeetCode/117-Populating-Next-Right-Pointers-in-Each-Node-II(Medium)/Solution2.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() {}

    Node(int _val, Node* _left, Node* _right, Node* _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
O(1) space, take advantage of next chain
*/
class Solution {
public:
    Node* connect(Node* root) {
        if (root == NULL) return root;
        
        Node* prev;
        Node* curr = new Node();
        
        prev = root;
        prev->next = NULL;

        curr->next = root;
        
        while(curr->next) {
            curr->next = NULL;
            
            Node* pPrev = prev;
            Node* pCurr = curr;
            while (pPrev) {
                if (pPrev->left) {
                    pCurr->next = pPrev->left;
                    pCurr = pCurr->next;
                }
                if (pPrev->right) {
                    pCurr->next = pPrev->right;
                    pCurr = pCurr->next;
                }

                pPrev = pPrev->next;
            }
            
            prev = curr->next;
        }
        
        return root;
    }
};
